README.md
PoshRSJob 1.7.4.4

[![Build status](https://ci.appveyor.com/api/projects/status/svrd4ho4otugki24?svg=true)](https://ci.appveyor.com/project/proxb/poshrsjob) [![Join the chat at https://gitter.im/proxb/PoshRSJob](https://badges.gitter.im/proxb/PoshRSJob.svg)](https://gitter.im/proxb/PoshRSJob?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Provides an alternative to PSjobs with greater performance and less overhead to run commands in the background, freeing up the console.

### Be sure to check out and contribute to the [Wiki](https://github.com/proxb/PoshRSJob/wiki)!

#### Download and install PoshRSJob from the PowerShellGallery using PowerShell:
```PowerShell
Install-Module -Name PoshRSJob
```

#### Download the latest release (1.7.4.4)
https://github.com/proxb/PoshRSJob/releases/download/1.7.4.4/PoshRSJob.zip


More information and examples here: http://learn-powershell.net/2015/04/19/latest-updates-to-poshrsjob/

Older post with some legacy examples found here: http://learn-powershell.net/2015/03/31/introducing-poshrsjob-as-an-alternative-to-powershell-jobs/

#### Now working on Linux/MacOS with PowerShell Core!
![alt tag](https://github.com/proxb/PoshRSJob/blob/master/Images/PoshRSJob.gif)

#### Examples
=================
```PowerShell
$Test = 'test'
$Something = 1..10
1..5|start-rsjob -Name {$_} -ScriptBlock {
        [pscustomobject]@{
            Result=($_*2)
            Test=$Using:Test
            Something=$Using:Something
        }
}            
Get-RSjob | Receive-RSJob
```
![alt tag](https://github.com/proxb/PoshRSJob/blob/master/Images/GetRSJob-ReceiveRSJob.gif)

#### This shows the streaming aspect with Wait-RSJob
```PowerShell
1..10|Start-RSJob {
    if (1 -BAND $_){
        "First ($_)"
    }Else{
        Start-sleep -seconds 2
        "Last ($_)"
    }
}|Wait-RSJob|Receive-RSJob|ForEach{"I am $($_)"}
```
![alt tag](https://github.com/proxb/PoshRSJob/blob/master/Images/RSJobStreamingExample.gif)

#### Nano Support for PoshRSJob
![alt tag](https://github.com/proxb/PoshRSJob/blob/master/Images/NanoPoshRSJob.png)


ReleaseNotes.md
---------
|1.7.4.4|
---------
* #175 Save and restore $env:PSModulePath until we find why it's being clobbered on PS Core

---------
|1.7.4.3|
---------
* #48 Moved garbage collection in runspacepool cleanup
* #48 Removed unnecessary Close() on runspacepool as Dispose() already in use
* #157 (Start-RSJob - be able to import private functions)
* #174 (Minor variable declaration)
* Variables will now always use local scope in runspace by default; not global which prevents incorrect data being presented if a variable is created and not cleaned up in each runspace.

---------
|1.7.3.11|
---------
* #139 (Wait-RSJob is not working with parameter set Name or ID)
* #140 (Wait-RSJob doesn't sleep)
* #141 (-RSJob question)

---------
|1.7.3.10|
---------
* #139 (Wait-RSJob is not working with parameter set Name or ID.)

---------
|1.7.3.9|
---------
* #42 (V2 Pester test hangs in AppVeyor)
* #121 (Missing output from Start-RSJob *bug*)
* #122 (Output stream is not always cleared when using pooled runspaces)
* #136 (RSJob on PSv2)

---------
|1.7.3.8|
---------
* #133 (Passing synchronized hashtable as argument does not work)
* #132 (Not handling single parameter in ArgumentList correctly)
* #136 (RSJob on PSv2) - Some parts may be fixed with how reflection is used

---------
|1.7.3.7|
---------
* #127 (Module loading failure with high throttle counts (and possibly related skipped jobs))
* #107 (Write-Stream throws lots of spurious errors on missing variables )

---------
|1.7.3.6|
---------
* (FunctionsToLoad doesn't work if a script is called within a script)
* #124 (powershell_ise processes are left around once PoshRSJob module is loaded)

---------
|1.7.3.5|
---------
* #119 (Passing an empty batch to Wait-RSJob stops all further processing in the caller)
* #95 (RSJob State does not reflect actual state of job)
* #111 (Exception calling "BeginInvoke" with "0" argument(s): "Cannot perform the operation because the runspace pool is not in the 'Opened' state. The current state is 'Closed'." )

---------
|1.7.3.3|
---------
* Fixed Issue #116 (v1.7.3.0 is extremely slow)
* Fixed Issue #75 (Feature Request: Add RunspaceID handling to Start-RSJob for better throttling support)
* Fixed Issue #107 (Write-Stream throws lots of spurious errors on missing variables)
* Added some better support for streams with Receive-RSJob

---------
|1.7.3.0|
---------
* Fixed Issue #112 (TabExpansion puts a small error in $error)
* Fixed Issue #115 (Multiple runspaces are being created when running Start-RSJob)

---------
|1.7.2.9|
---------
* Fixed Issue #101 (Wait-RsJob -State Completed with no input returns Attempted to divide by zero.)
* Fixed Issue #108 (Caveats of Start-RSJob -ModulesToImport) - Using ForEach loop with PoshRSJob no longer works with this update due to issues that it brought with -ModulesToImport, -FunctionsToLoad and -PSSnapinsToImport where these would fail due to a runspacepool already being used.

---------
|1.7.2.7|
---------
* Fixed Issue #102 (Receive-RsJob doesn't process -InputObject properly)

---------
|1.7.2.6|
---------
* Fixed Issue #99 (Add an InputObject property to RSJob)

---------
|1.7.2.5|
---------
* Fixed Issue #96 (Error when -FunctionsToLoad parameter is used and the Function does not have an alias)

---------
|1.7.2.4|
---------
* Fixed Issue #92 (Cannot load module in PS4 due to "class" keyword)


---------
|1.7.2.3|
---------
* Fixed Issue #87 (Stop-RSJob gives an error if it has no input)

---------
|1.7.2.2|
---------
* Fixed Issue #59 (Receive-RSJob doesn't clear a job's HasMoreData state)

---------
|1.7.2.1|
---------
* Fixed Issue #83 (FunctionsToImport should include the function's Alias where applicable)


---------
|1.7.1.0|
---------
* Replaced private apis with public apis (#85 Update RunspaceConfiguration apis to use InitialSessionState instead)

---------
|1.7.0.0|
---------
* Remove need for DLL file for building out the classes. Using pure PowerShell (mostly) via means of here-strings and Add-Type for PowerShell V2-4 and the new Classes keywords for PowerShell V5 which includes PowerShell Core/Nano.
* Remove the prefixes for custom objects so they no longer start with PoshRS.PowerShell.. Now they are V2UsingVariable, RSJob and RSRunspacePool.

---------
|1.6.2.1|
---------
* Add support for PowerShell Core on Linux/MacOS (this still needs more work but should load types within a runspace now!)

---------
|1.6.1.0|
---------
* Fixed Issue #75 (Feature Request: Add RunspaceID handling to Start-RSJob for better throttling support)
* Fixed Issue #82 (Exception setting "RunspacePool" in 1.6.0.0 build)

---------
|1.5.7.7|
---------
* Fixed Issue #69 (Module produces error if imported more than once (PS v2 only))
* Fixed Issue #64 (HadErrors in PoshRS.PowerShell.RSJob throws errors in PowerShell V2)
* Fixed Issue #67 (Converted Add-Type code for C# classes to be created via Reflection for Nano server support) <- Created custom dll
* Fixed Issue #61 (Receive-RSJob not allowing -Job parameter input)
* Fixed Issue #63 (Replaced Global variables with Script scope)
* Fixed Issue #66 (Parameters don't work with PowerShell V2)
* Fixed Issue #65 (Bug with v2 variable substitution - single-quoted strings get $var references replaced)
* Fixed Issue #68 (ApartmentState Does Not Exist in Nano)
* Fixed Issue #76 (Jobs don't have output when using ADSI WinNT provider (Receive-RSJob))


Scripts/Start-RSJob.ps1
Function Start-RSJob {
    <#
        .SYNOPSIS
            Starts a background job using runspaces.

        .DESCRIPTION
            This will run a command in the background, leaving your console available to perform other tasks. This uses
            runspaces in runspacepools which allows for throttling of running jobs. As the jobs are finished, they will automatically
            dispose of each runspace and allow other runspace jobs in queue to begin based on the throttling of the runspacepool.

            This is available on PowerShell V3 and above. By doing this, you can use the $Using: variable to take variables
            in the local scope and apply those directly into the scriptblock of the background runspace job.

        .PARAMETER ScriptBlock
            The scriptblock that holds all of the commands which will be run in the background runspace. You must specify
            at least one Parameter in the Param() to account for the item that is being piped into Start-Job.

        .PARAMETER FilePath
            This is the path to a file containing code that will be run in the background runspace job.

        .PARAMETER InputObject
            The object being piped into Start-RSJob or applied via the parameter.

        .PARAMETER Name
            The name of a background runspace job

        .PARAMETER ArgumentList
            List of values that will be applied at the end of the argument list in the Param() statement.

        .PARAMETER Throttle
            Number of concurrent running runspace jobs which are allowed at a time.

        .PARAMETER ModulesToImport
            A collection of modules that will be imported into the background runspace job.

        .PARAMETER PSSnapinsToImport
            A collection of PSSnapins that will be imported into the background runspace job.

        .PARAMETER FunctionsToImport
            A collection of functions that will be imported for use with a background runspace job.

        .NOTES
            Name: Start-RSJob
            Author: Boe Prox                

        .EXAMPLE
            Get-ChildItem -Directory | Start-RSjob -Name {$_.Name} -ScriptBlock {
                Param($Directory)
                Write-Verbose $_
                $Sum = (Get-ChildItem $Directory.FullName -Recurse -Force -ErrorAction SilentlyContinue |
                Measure-Object -Property Length -Sum).Sum
                [pscustomobject]@{
                    Name = $Directory.Name
                    SizeMB = ([math]::round(($Sum/1MB),2))
                }
            } 
            
            Id  Name                 State           HasMoreData  HasErrors    Command
            --  ----                 -----           -----------  ---------    -------
            11  .shsh                Running         False        False        ...
            12  .ssh                 Running         False        False        ...
            13  Contacts             Running         False        False        ...
            14  Desktop              Running         False        False        ...
            15  Documents            Running         False        False        ...
            16  Downloads            Running         False        False        ...
            17  Favorites            Running         False        False        ...
            18  Links                Running         False        False        ...
            19  Music                Running         False        False        ...
            20  OneDrive             Running         False        False        ...
            21  Pictures             Running         False        False        ...
            22  Saved Games          Running         False        False        ...
            23  Searches             Running         False        False        ...
            24  Videos               Running         False        False        ...

            Get-RSJob | Receive-RSJob

            Name          SizeMB
            ----          ------
            .shsh              0
            .ssh               0
            Contacts           0
            Desktop         7.24
            Documents      83.99
            Downloads    10259.6
            Favorites          0
            Links              0
            Music       16691.89
            OneDrive     1485.24
            Pictures     1734.91
            Saved Games        0
            Searches           0
            Videos         17.19

            Description
            -----------
            Starts a background runspace job that looks at the total size of each folder. Using Get-RSJob | Recieve-RSJob shows 
            the results when the State is Completed.         

        .EXAMPLE
            $Test = 'test'
            $Something = 1..10
            1..5|start-rsjob -Name {$_} -ScriptBlock {
                Param($Object) [pscustomobject]@{
                    Result=($Object*2)
                    Test=$Using:Test
                    Something=$Using:Something
                }
            }            

            Id  Name                 State           HasMoreData  HasErrors    Command
            --  ----                 -----           -----------  ---------    -------
            76  1                    Completed       True         False        ...
            77  2                    Running         False        False        ...
            78  3                    Running         False        False        ...
            79  4                    Completed       False        False        ...
            80  5                    Completed       False        False        ...
            
            Get-RSjob | Receive-RSJob

            Result Test Something
            ------ ---- ---------
                 2 test {1, 2, 3, 4...}
                 4 test {1, 2, 3, 4...}
                 6 test {1, 2, 3, 4...}
                 8 test {1, 2, 3, 4...}
                10 test {1, 2, 3, 4...}
            
            Description
            -----------
            Shows an example of the $Using: variable being used in the scriptblock.

        .EXAMPLE
            $Test = 42
            $AnotherTest = 7
            $String = 'SomeString'
            $ProcName = 'powershell_ise'
            $ScriptBlock = {
                Param($y,$z)
                [pscustomobject] @{
                    Test = $y
                    Proc = (Get-Process -Name $Using:ProcName)
                    String = $Using:String
                    AnotherTest = ($z+$_)
                    PipedObject = $_
                }
            }

            1..5|Start-RSJob $ScriptBlock -ArgumentList $test, $anothertest
            
            Description
            -----------
            Shows an example of the $Using: variable being used in the scriptblock as well as $_ and multiple -ArgumentList parameters.

    #>
    [OutputType('PoshRS.PowerShell.RSJob')]
    [cmdletbinding(
        DefaultParameterSetName = 'ScriptBlock'
    )]
    Param (
        [parameter(Position=0,ParameterSetName = 'ScriptBlock')]
        [ScriptBlock]$ScriptBlock,
        [parameter(Position=0,ParameterSetName = 'ScriptPath')]
        [string]$FilePath,
        [parameter(ValueFromPipeline=$True,ValueFromPipelineByPropertyName=$True)]
        [object]$InputObject,
        [parameter()]
        [object]$Name,
        [parameter()]
        [object]$ArgumentList,
        [parameter()]
        [int]$Throttle = 5,
        [parameter()]
        [string[]]$ModulesToImport,
        [parameter()]
        [string[]]$PSSnapinsToImport,
        [parameter()]
        [string[]]$FunctionsToLoad
    )
    Begin {  
        If ($PSBoundParameters['Debug']) {
            $DebugPreference = 'Continue'
        } 
        If ($PSBoundParameters.ContainsKey('Name')) {
            If ($Name -isnot [scriptblock]) {
                $JobName = [scriptblock]::Create("Write-Output $Name")
            } Else {
                $JobName = [scriptblock]::Create( ($Name -replace '\$_','$Item'))
            }
        } Else {
            Write-Verbose "Creating default Job Name"
            $JobName = [scriptblock]::Create('Write-Output Job$($Id)')
        }
        $RunspacePoolID = [guid]::NewGuid().ToString()
        Write-Verbose "Creating runspacepool <$($RunspacePoolID)> with max threads: $Throttle"        
        $InitialSessionState = [System.Management.Automation.Runspaces.InitialSessionState]::CreateDefault()
        If ($PSBoundParameters['ModulesToImport']) {
            [void]$InitialSessionState.ImportPSModule($ModulesToImport)
        }
        If ($PSBoundParameters['PSSnapinsToImport']) {
            [void]$InitialSessionState.ImportPSSnapIn($PSSnapinsToImport,[ref]$Null)
        }
        If ($PSBoundParameters['FunctionsToLoad']) {
            ForEach ($Function in $FunctionsToLoad) {
                Try {
                    $Definition = Get-Content Function:\$Function -ErrorAction Stop
                    $SessionStateFunction = New-Object System.Management.Automation.Runspaces.SessionStateFunctionEntry -ArgumentList $Function, $Definition
                    $InitialSessionState.Commands.Add($SessionStateFunction) 
                } Catch {
                    Write-Warning "$($Function): $($_.Exception.Message)"
                }
            }           
        }
        $RunspacePool = [runspacefactory]::CreateRunspacePool($InitialSessionState)
        [void]$RunspacePool.SetMaxRunspaces($Throttle)
        If ($PSVersionTable.PSVersion.Major -gt 2) {
            $RunspacePool.CleanupInterval = [timespan]::FromMinutes(2)    
        }
        $RunspacePool.Open()
        If ($PSBoundParameters['FilePath']) {
            $ScriptBlock = [scriptblock]::Create((Get-Content $FilePath | Out-String))
        }
        $RSPObject = New-Object PoshRS.PowerShell.RSRunspacePool -Property @{
            RunspacePool = $RunspacePool
            MaxJobs = $RunspacePool.GetMaxRunspaces()
            RunspacePoolID = $RunspacePoolID
        }
        $Script:List = New-Object System.Collections.ArrayList
        If ($PSBoundParameters.ContainsKey('InputObject')) {            
            [void]$list.AddRange($InputObject)    
            $IsPipeline = $False
        } Else {
            $IsPipeline = $True
        }
    }
    Process {
        If ($IsPipeline -AND $PSBoundParameters.ContainsKey('InputObject')) {
            [void]$List.Add($InputObject)
        }
    }
    End {  
        $SBParamCount = @(GetParamVariable -ScriptBlock $ScriptBlock).Count
        If ($PSBoundParameters.ContainsKey('InputObject')) {
            $SBParamCount++
        }
        If ($ArgumentList.Count -ne $SBParamCount) {
            Write-Verbose 'Will use $_ in Param() Block'
            $Script:Add_ = $True
        } Else {
            $Script:Add_ = $False
        }
        #Convert ScriptBlock for $Using:
        Switch ($PSVersionTable.PSVersion.Major) {
            2 {
                Write-Debug "Using PSParser with PowerShell V2"
                $UsingVariables = @(GetUsingVariablesV2 -ScriptBlock $ScriptBlock)
                
                If ($UsingVariables.count -gt 0) {
                    $UsingVariableValues = @(GetUsingVariableValuesV2 -UsingVar $UsingVariables)
                    Write-Verbose ("Found {0} `$Using: variables!" -f $UsingVariableValues.count)
                }
                If ($UsingVariables.count -gt 0 -OR $Script:Add_) {
                    $NewScriptBlock = ConvertScriptBlockV2 $ScriptBlock            
                } Else {
                    $NewScriptBlock = $ScriptBlock
                }                
            }
            Default {
                Write-Debug "Using AST with PowerShell V3+"
                $UsingVariables = @(GetUsingVariables $ScriptBlock | Group SubExpression | ForEach {
                    $_.Group | Select -First 1
                })
                If ($UsingVariables.count -gt 0) {
                    $UsingVariableValues = @(GetUsingVariableValues -UsingVar $UsingVariables)
                    Write-Verbose ("Found {0} `$Using: variables!" -f $UsingVariableValues.count)
                }
                If ($UsingVariables.count -gt 0 -OR $Script:Add_) {
                    $NewScriptBlock = ConvertScript $ScriptBlock            
                } Else {
                    $NewScriptBlock = $ScriptBlock
                }
            }
        }

        Write-Debug "ScriptBlock: $($NewScriptBlock)"
        [System.Threading.Monitor]::Enter($RunspacePools.syncroot) 
        [void]$RunspacePools.Add($RSPObject)
        [System.Threading.Monitor]::Exit($RunspacePools.syncroot) 

        If ($List.Count -gt 0) {
            Write-Debug "InputObject"
            ForEach ($Item in $list) {
                $ID = Increment                    
                Write-Verbose "Using $($Item) as pipline variable"
                $PowerShell = [powershell]::Create().AddScript($NewScriptBlock)
                $PowerShell.RunspacePool = $RunspacePool
                [void]$PowerShell.AddArgument($Item)
                If ($UsingVariableValues.count -gt 0) {
                    For ($i=0;$i -lt $UsingVariableValues.count;$i++) {
                        Write-Verbose "Adding Param: $($UsingVariableValues[$i].Name) Value: $($UsingVariableValues[$i].Value)"
                        [void]$PowerShell.AddParameter($UsingVariableValues[$i].NewVarName,$UsingVariableValues[$i].Value)
                    }
                }
                If ($PSBoundParameters.ContainsKey('ArgumentList')) {
                    ForEach ($Argument in $ArgumentList) {
                        Write-Verbose "Adding Argument: $($Argument)"
                        [void]$PowerShell.AddArgument($Argument)    
                    }
                }
                $Handle = $PowerShell.BeginInvoke()
                $_JobName = If ($PSVersionTable.PSVersion.Major -eq 2) {
                    $JobName.Invoke()
                } Else {
                    $JobName.InvokeReturnAsIs()
                }
                $Object = New-Object PoshRS.PowerShell.RSJob -Property @{
                    Name = $_JobName
                    InstanceID = [guid]::NewGuid().ToString()
                    ID = $ID  
                    Handle = $Handle
                    InnerJob = $PowerShell
                    Runspace = $PowerShell.Runspace
                    Finished = $handle.IsCompleted
                    Command  = $ScriptBlock.ToString()
                    RunspacePoolID = $RunSpacePoolID
                    State = [System.Management.Automation.PSInvocationState]::Running
                }
                
                $RSPObject.LastActivity = Get-Date
                [System.Threading.Monitor]::Enter($Jobs.syncroot) 
                [void]$Jobs.Add($Object)
                [System.Threading.Monitor]::Exit($Jobs.syncroot) 
                $Object            
            }
        } Else {
            Write-Debug "No InputObject"
            $ID = Increment                    
            $PowerShell = [powershell]::Create().AddScript($NewScriptBlock)
            $PowerShell.RunspacePool = $RunspacePool
            If ($UsingVariableValues) {
                For ($i=0;$i -lt $UsingVariableValues.count;$i++) {
                    Write-Verbose "Adding Param: $($UsingVariableValues[$i].Name) Value: $($UsingVariableValues[$i].Value)"
                    [void]$PowerShell.AddParameter($UsingVariableValues[$i].NewVarName,$UsingVariableValues[$i].Value)
                }
            }
            If ($PSBoundParameters.ContainsKey('ArgumentList')) {
                ForEach ($Argument in $ArgumentList) {
                    Write-Verbose "Adding Argument: $($Argument)"
                    [void]$PowerShell.AddArgument($Argument)    
                }
            }
            $Handle = $PowerShell.BeginInvoke()
            $_JobName = If ($PSVersionTable.PSVersion.Major -eq 2) {
                $JobName.Invoke()
            } Else {
                $JobName.InvokeReturnAsIs()
            }
            $Object = New-Object PoshRS.PowerShell.RSJob -Property @{
                Name = $_JobName
                InstanceID = [guid]::NewGuid().ToString()
                ID = $ID  
                Handle = $Handle
                InnerJob = $PowerShell
                Runspace = $PowerShell.Runspace
                Finished = $handle.RSWaitHandle
                Command  = $ScriptBlock.ToString()
                RunspacePoolID = $RunSpacePoolID
            }
            
            $RSPObject.LastActivity = Get-Date
            [System.Threading.Monitor]::Enter($Jobs.syncroot) 
            [void]$Jobs.Add($Object)
            [System.Threading.Monitor]::Exit($Jobs.syncroot) 
            $Object        
        }      
    }
}
